internal/webhook/package_validator.go

```diff
@@ -0,0 +1,129 @@
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+type PackageValidatingWebhook struct {
+	client.Client
+	*owners.OwnerManager
+	*dependency.DependendcyManager
+	repo repoclient.RepoClient
+}
+
+//+kubebuilder:webhook:path=/validate-packages-glasskube-dev-v1alpha1-package,mutating=false,failurePolicy=fail,sideEffects=None,groups=packages.glasskube.dev,resources=packages,verbs=create;update;delete,versions=v1alpha1,name=vpackage.kb.io,admissionReviewVersions=v1
+
+var _ webhook.CustomValidator = &PackageValidatingWebhook{}
+
+func (p *PackageValidatingWebhook) SetupWithManager(mgr ctrl.Manager) error {
+	p.OwnerManager = owners.NewOwnerManager(p.Scheme())
+	p.DependendcyManager = dependency.NewDependencyManager(ctrladapter.NewControllerRuntimeAdapter(p.Client))
+	p.repo = repo.DefaultClient
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(&v1alpha1.Package{}).
+		WithValidator(p).
+		Complete()
+}
+
+// ValidateCreate implements admission.CustomValidator.
+func (p *PackageValidatingWebhook) ValidateCreate(ctx context.Context, obj runtime.Object) (warnings admission.Warnings, err error) {
+	log := ctrl.LoggerFrom(ctx)
+	if pkg, ok := obj.(*v1alpha1.Package); ok {
+		log.Info("validate create", "name", pkg.Name)
+		return nil, p.validateCreateOrUpdate(ctx, pkg)
+	}
+	return nil, ErrInvalidObject
+}
+
+// ValidateUpdate implements admission.CustomValidator.
+func (p *PackageValidatingWebhook) ValidateUpdate(ctx context.Context, oldObj runtime.Object, newObj runtime.Object) (warnings admission.Warnings, err error) {
+	log := ctrl.LoggerFrom(ctx)
+	if oldPkg, ok := oldObj.(*v1alpha1.Package); ok {
+		if newPkg, ok := newObj.(*v1alpha1.Package); ok {
+			if oldPkg.Spec.PackageInfo == newPkg.Spec.PackageInfo {
+				// If the package info did not change, we are already done
+				return nil, nil
+			} else {
+				log.Info("validate update", "name", newPkg.Name)
+				return nil, p.validateCreateOrUpdate(ctx, newPkg)
+			}
+		}
+	}
+	return nil, ErrInvalidObject
+}
+
+// ValidateDelete implements admission.CustomValidator.
+//
+// A package can not be deleted if another package depends on it. This is the case if it has owner references
+// to at least one other package. However, we do have to allow packages to be deleted if all their dependant
+// packages are already being deleted, because the webhook is also called if a package is garbage-collected and
+// this would lead to a dead lock otherwise:
+// The dependant package is not deleted because of the foreground deletion finalizer and the dependency is not
+// deleted because the dependant is not deleted.
+func (p *PackageValidatingWebhook) ValidateDelete(ctx context.Context, obj runtime.Object) (warnings admission.Warnings, err error) {
+	log := ctrl.LoggerFrom(ctx)
+	if pkg, ok := obj.(*v1alpha1.Package); ok {
+		log.Info("validate delete", "name", pkg.Name)
+		if dependants, err := p.OwnersOfType(&v1alpha1.Package{}, pkg); err != nil {
```

Comment:
I think this is missing the case in which there is a package dependent on `pkg` without having an owner reference on it.

Reply:
This is a good point. As discussed, we will implement this in a separate PR, as a similar functionality is also needed in other places.